core/src/main/java/org/apache/iceberg/avro/NameMappingWithAvroSchema.java

```diff
@@ -0,0 +1,123 @@
+import org.apache.iceberg.relocated.com.google.common.base.Preconditions;
+import org.apache.iceberg.relocated.com.google.common.collect.Lists;
+import org.apache.iceberg.types.Type;
+import org.apache.iceberg.types.Types;
+
+public class NameMappingWithAvroSchema extends AvroWithTypeByStructureVisitor<MappedFields> {
+  @Override
+  public MappedFields record(
+      Type struct, Schema record, List<String> names, List<MappedFields> fieldResults) {
+    List<MappedField> fields = Lists.newArrayListWithExpectedSize(fieldResults.size());
+
+    for (int i = 0; i < fieldResults.size(); i += 1) {
+      Types.NestedField field = struct.asStructType().fields().get(i);
+      MappedFields result = fieldResults.get(i);
+      fields.add(MappedField.of(field.fieldId(), field.name(), result));
+    }
+
+    return MappedFields.of(fields);
+  }
+
+  @Override
+  public MappedFields union(Type type, Schema union, List<MappedFields> optionResults) {
+    if (AvroSchemaUtil.isOptionSchema(union)) {
+      for (int i = 0; i < optionResults.size(); i += 1) {
+        if (union.getTypes().get(i).getType() != Schema.Type.NULL) {
+          return optionResults.get(i);
+        }
+      }
+    } else { // Complex union
+      Preconditions.checkArgument(
+          type instanceof Types.StructType,
+          "Cannot visit invalid Iceberg type: %s for Avro complex union type: %s",
+          type,
+          union);
+      Types.StructType struct = (Types.StructType) type;
+      List<MappedField> fields = Lists.newArrayListWithExpectedSize(optionResults.size());
+      int index = 0;
+      // Avro spec for union types states that unions may not contain more than one schema with the
+      // same type, except for the named types record, fixed and enum. For example, unions
+      // containing two array types or two map types are not permitted, but two types with different
+      // names are permitted.
+      // Therefore, for non-named types, use the Avro type toString() as the field mapping key. For
+      // named types, use the record name of the Avro type as the field mapping key.
+      for (Schema option : union.getTypes()) {
+        if (option.getType() != Schema.Type.NULL) {
+          // Check if current option is a named type, i.e., a RECORD, ENUM, or FIXED type. If so,
+          // use the record name of the Avro type as the field name. Otherwise, use the Avro
+          // type toString().
+          if (option.getType() == Schema.Type.RECORD
+              || option.getType() == Schema.Type.ENUM
+              || option.getType() == Schema.Type.FIXED) {
+            fields.add(
+                MappedField.of(
+                    struct.fields().get(index).fieldId(),
+                    option.getName(),
+                    optionResults.get(index)));
+          } else {
+            fields.add(
+                MappedField.of(
+                    struct.fields().get(index).fieldId(),
+                    option.toString(),
```

Comment:
I think this should be `option.getType().getName()` instead of `toString`. Using `toString` could cause problems when a type has extra metadata. We want to use the type name specifically, so we shouldn't rely on the schema's representation being the same as the type name.